WordPress/src/main/java/org/wordpress/android/widgets/WPLoginInputRow.java

```diff
@@ -0,0 +1,236 @@
+        }
+        mIcon.setLayoutParams(iconLayoutParams);
+
+        RelativeLayout.LayoutParams editTextLayoutParams = (LayoutParams) mTextInputLayout.getLayoutParams();
+        rules = editTextLayoutParams.getRules();
+        for (int i = 0; i < rules.length; i++) {
+            if (rules[i] == mIcon.getId()) {
+                if (Build.VERSION.SDK_INT > 23) {
+                    rules[i] = mNewIds.get(0);
+                } else {
+                    editTextLayoutParams.addRule(i, mNewIds.get(0));
+                }
+            }
+        }
+        mTextInputLayout.setLayoutParams(editTextLayoutParams);
+
+        mIcon.setId(mNewIds.get(0));
+        mTextInputLayout.setId(mNewIds.get(1));
+        mEditText.setId(mNewIds.get(2));
+    }
+
+    @Override
+    protected Parcelable onSaveInstanceState() {
+        Parcelable superState = super.onSaveInstanceState();
+        return new SavedState(superState, mNewIds);
+    }
+
+    @Override
+    protected void onRestoreInstanceState(Parcelable state) {
+        //begin boilerplate code so parent classes can restore state
+        if(!(state instanceof SavedState)) {
+            super.onRestoreInstanceState(state);
+            return;
+        }
+
+        SavedState savedState = (SavedState) state;
+        super.onRestoreInstanceState(savedState.getSuperState());
+
+        mNewIds = savedState.mIds;
+
+        reassignIds();
+    }
+
+    public void addTextChangedListener(TextWatcher watcher) {
+        mEditText.addTextChangedListener(watcher);
+    }
+
+    public void setOnEditorCommitListener(final OnEditorCommitListener listener) {
+        mEditText.setOnEditorActionListener(new TextView.OnEditorActionListener() {
+            @Override
+            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
+                if (actionId == EditorInfo.IME_ACTION_DONE ||
+                        actionId == EditorInfo.IME_ACTION_NEXT ||
+                        (event != null
+                                && event.getAction() == KeyEvent.ACTION_UP
+                                && event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) {
+                    listener.OnEditorCommit();
+                }
+
+                // always consume the event so the focus stays in the EditText
+                return true;
+            }
+        });
+    }
+
+    public void setOnEditorActionListener(TextView.OnEditorActionListener l) {
+        mEditText.setOnEditorActionListener(l);
+    }
+
+    public final void setText(CharSequence text) {
+        mEditText.setText(text);
+    }
+
+    public void setError(@Nullable final CharSequence error) {
+        mTextInputLayout.setError(error);
+    }
+
+    private static class SavedState extends BaseSavedState {
+        private List<Integer> mIds;
+
+        SavedState(Parcelable superState, List<Integer> ids) {
+            super(superState);
+            mIds = ids;
+        }
+
+        private SavedState(Parcel in) {
```

Comment:
This `SavedState` constructor can be made package-private to avoid accessing a private member in the `createFromParcel` method.